### PR TITLE
By yanniboi: Allow sorting of dashboard in manage mode.

### PR DESCRIPTION
**Pull request by yanniboi**
Test site: http://nw.contacts.freelygive.org.uk/admin/contacts/2

Theme merge:
https://github.com/FreelyGive/contacts_theme/pull/10

### CODE REVIEW

**Comment by yanniboi** [comment]
selection_275

File: js/dashboard.manage.js [hunk 1]
```diff
@@ -1,17 +1,108 @@
...
-        url = '/admin/contacts/ajax/manage-off-canvas/'+tab+'/'+name+'?'+destination;
-    $block.addClass('manage-wrapper').prepend(Drupal.theme('manageTrigger', url));
+    var tab = $block.data('contacts-manage-block-tab'),
+      name = $block.data('contacts-manage-block-name'),
+      url = [['/admin/contacts/ajax/manage-off-canvas/', tab, name].join('/'), destination].join('?');
+
+    $block.addClass('manage-wrapper');
+    var link = $block.find('.manage-trigger');
+    if (link.length !== 0) {
+      link.find('a').attr('data-ajax-url', url);
```
Comment (andrewbelcher) [inline js/dashboard.manage.js]: If the link already exists, can we build this as part of rendering but keep it hidden?

File: js/dashboard.manage.js [hunk 2]
```diff
@@ -1,17 +1,108 @@
...
+    for (var weight = 0; weight < ids.length; weight++) {
+      var el = $('[data-contacts-manage-block-name=' + ids[weight] + ']');
+
+      // @todo check that profile type and relationship are available.
+      var block_data = {
+        name: ids[weight],
+        id: el.data('contacts-manage-block-id'),
+        entity_type: el.data('contacts-manage-entity-type'),
+        entity_bundle: el.data('contacts-manage-entity-bundle'),
+        entity_relationship: el.data('contacts-manage-entity-relationship')
```
Comment (andrewbelcher) [inline js/dashboard.manage.js]: Presumably the server already knows what plugin id, entity type etc a block is for, so can we not just use name?

File: modules/crm_tools/js/open-iconic.ajax.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+(function ($, Drupal) {
+
+  Drupal.behaviors.openIconicSpriteAjaxLoad = {
+    attach: function attach(context) {
+      $.get("/modules/contrib/contacts/modules/crm_tools/includes/open-iconic.svg", function(data) {
```
Comment (andrewbelcher) [inline modules/crm_tools/js/open-iconic.ajax.js]: This should be stored in `drupalSetting` as we don't know the base path.

File: modules/crm_tools/templates/crm-tools-hat.html.twig [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<span {{ attributes.setAttribute('title', label) }}>
+    {{ label }}
+</span>
```
Comment (andrewbelcher) [inline modules/crm_tools/templates/crm-tools-hat.html.twig]: Missing new line at EOF

File: src/Plugin/Block/ContactsEntity.php [hunk 5]
```diff
@@ -205,14 +205,23 @@ public function getEditLink($mode) {
...
-    $variables['attributes']['data-contacts-manage-entity-bundle'] = $variables['bundle'] = $definition['_bundle_id'];
-
-    $variables['footer']['links'] = [
+    $variables['content']['links'] = [
       '#theme' => 'item_list',
       '#list_type' => 'ul',
       '#items' => $this->getManageLinks(),
     ];
+
+    if (!empty($definition['_required_hats'])) {
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsEntity.php]: Should this be on `ContactsEntity`, or should this be part of the manage mode code as it's not block specific?

**Review by andrewbelcher — CHANGES_REQUESTED** [review]

File: src/Plugin/Block/ContactsEntity.php (hunk 5 above)
Comment (yanniboi) [reply src/Plugin/Block/ContactsEntity.php]: hmmm... No I think this is the right place for it. ContactsEntity is a dashboard block, and blocks can decide they require certain hats. So far its just profiles that do that, but I can image other scenarios.

File: js/dashboard.manage.js (hunk 1 above)
Comment (yanniboi) [reply js/dashboard.manage.js]: I think this is quite tricky to solve now. Can we do a followup for this? I think we'll need to do some ajax link involved cleanup at some point anyway.

Comment (andrewbelcher) [reply js/dashboard.manage.js]: Yup - put an `@todo` in linking to a follow up issue and that's me happy!

File: src/Plugin/Block/ContactsEntity.php (hunk 5 above)
Comment (andrewbelcher) [reply src/Plugin/Block/ContactsEntity.php]: So although `ContactsEntity` is the only block that currently supports it, putting it here means that other blocks _can't_ support it. Having said that - not a priority for now, so let's put an `@todo` in with a link to an issue to decide/change this.

**Review by andrewbelcher — APPROVED** [review]
Changes look good, let's just get those two `@todo`s in and feel free to merge!

**Pull request merged**